include/deal.II/matrix_free/fe_evaluation_data.h

```diff
@@ -371,41 +399,199 @@ class FEEvaluationBaseData
+             "faces. Use get_face_no() in other cases."));
+
+    return all_face_numbers;
+  }
+
+  /**
+   * Store the orientation of the neighbor's faces with respect to the current
+   * cell for the case of exterior faces on ECL with possibly different
+   * orientations behind different cells.
+   *
+   * @note Only available for `dof_access_index == dof_access_cell` and
+   * `is_interior_face == false`.
+   */
+  const std::array<std::uint8_t, n_lanes> &
+  get_all_face_orientations() const
+  {
+    // implemented inline to avoid compilation problems on Windows
+    Assert(cell != numbers::invalid_unsigned_int, ExcNotInitialized());
+    Assert(is_face &&
+             dof_access_index ==
+               internal::MatrixFreeFunctions::DoFInfo::dof_access_cell &&
+             is_interior_face == false,
+           ExcMessage(
+             "All face numbers can only be queried for ECL at exterior "
+             "faces. Use get_face_no() in other cases."));
+
+    return all_face_orientations;
+  }
+
   //@}
 
-protected:
   /**
-   * Constructor. Made protected to prevent users from directly using this
-   * class. Takes all data stored in MatrixFree. If applied to problems with
-   * more than one quadrature formula selected during construction of
-   * `matrix_free`, `quad_no` allows to select the appropriate formula.
+   * This data structure is used for the initialization by the derived
+   * FEEvaluation classes.
    */
-  FEEvaluationBaseData(const std::tuple<const ShapeInfoType *,
-                                        const DoFInfo *,
-                                        unsigned int,
-                                        unsigned int> &info,
-                       const MappingInfoStorageType &  mapping_data,
-                       const unsigned int              quad_no,
-                       const bool                      is_interior_face,
-                       const unsigned int              face_type);
+  struct InitializationData
+  {
+    const ShapeInfoType *         shape_info;
+    const DoFInfo *               dof_info;
+    const MappingInfoStorageType *mapping_data;
+    unsigned int                  active_fe_index;
+    unsigned int                  active_quad_index;
+    const typename MappingInfoStorageType::QuadratureDescriptor *descriptor;
+  };
 
+protected:
   /**
-   * Constructor that comes with reduced functionality and works similar as
-   * FEValues.
+   * Constructor filling information available from a MatrixFree class,
+   * collected in an internal data structure to reduce overhead of
+   * initialization. Used in derived classes when this class is initialized
+   * from a MatrixFree object.
    */
-  FEEvaluationBaseData(
-    const std::shared_ptr<
-      internal::MatrixFreeFunctions::
-        MappingDataOnTheFly<dim, Number, VectorizedArrayType>> &mapping_data);
+  FEEvaluationData(const InitializationData &info,
```

Comment:
Name `info` does not really fit to class name in my opinion, why not `init_data` for example?